### PR TITLE
nwchem: add new package

### PR DESCRIPTION
**Pull request by davydden**

### CODE REVIEW

File: var/spack/repos/builtin/packages/nwchem/package.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    version('6.6', 'c581001c004ea5e5dfacb783385825e3',
+            url='http://www.nwchem-sw.org/images/Nwchem-6.6.revision27746-src.2015-10-20.tar.gz')
+
+    depends_on('blas')
+    depends_on('lapack')
+    depends_on('mpi')
+
+    # build
+    # TODO: tries to fetch python 2.7, instead of using 2.7.11
+    # depends_on('python@2.7', type='build')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nwchem/package.py]: This is a known bug. See #1035. A workaround for now is to use `depends_on('python@2.7:2.8')`

File: var/spack/repos/builtin/packages/nwchem/package.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    def install(self, spec, prefix):
+        # see http://www.nwchem-sw.org/index.php/Compiling_NWChem
+
+        args = []
+        args.extend([
+            'NWCHEM_TOP=%s' % self.stage.source_path,
+            'USE_MPI=y',
+            'MPI_LOC=%s' % spec['mpi'].prefix,
+            'USE_PYTHONCONFIG=y',
+            'PYTHONVERSION=2.7',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/nwchem/package.py]: Once you fix the depends_on, use `spec['python'].version.up_to(2)` here

**Comment by davydden** [comment]
don't merge yet, it looks like the package needs some more copying from the source...

**Comment by tgamblin** [comment]
@davydden: is this WIP or good to go?

**Comment by davydden** [comment]
@tgamblin it's good to go. There are some issues for macOS i hope to solve soon (see http://www.nwchem-sw.org/index.php/Special:AWCforum/st/id2138/#post_7513), but the current state compiles and runs fine on Ubuntu + GCC.

**Comment by tgamblin** [comment]
@davydden: do you want to fix the OS X stuff first or would you rather just merge?

**Comment by davydden** [comment]
@tgamblin i would merge as is. I hope to address OS-X issues in subsequent PR soon.

**Comment by tgamblin** [comment]
Thanks!

**Comment by alalazo** [comment]
Sorry if the question may sound dumb but are really 9k lines of patches necessary for this package ?

**Comment by davydden** [comment]
> Sorry if the question may sound dumb but are really 9k lines of patches necessary for this package ?

unfortunately yes. I suggested to developers to use maintenance versions http://www.nwchem-sw.org/index.php/Special:AWCforum/st/id2142/use_maintenance_versions_%286.6.1%2C....html , but i don't think this will happen

**Comment by davydden** [comment]
@alalazo p.s. patches could be removed once https://github.com/LLNL/spack/issues/1584 is implemented.

**Comment by jeffhammond** [comment]
It would be cool if Spack supported the latest development version of NWChem (https://github.com/nwchemgit/nwchem) in addition to releases.  Please also submit pull requests associated with your patches.

**Comment by davydden** [comment]
@jeffhammond all patches are provided by NWChem, the current state package downloads them directly, see
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/nwchem/package.py
As for development version, this should be easy to add, you want to give it a try?

**Comment by jeffhammond** [comment]
@davydden Creating pull requests based upon third-party patches is problematic on a number of levels.  The biggest issue is that I'd be signing my name to code that I have no way of knowing is clean from a license and IP perspective.  This is particularly relevant in this case since the size of the patches is 9000 lines, which is about 1000 times larger than the patch size that I might be able to reasonably claim is not significant enough to warrant copyright.  While these issues may not matter to others, my contributions to NWChem follow my employer's guidelines on open-source development.

On an engineering level, it isn't appropriate for me to put forth changes to NWChem that I don't fully understand and can't answer questions about.  The best way to ensure that pull requests are reviewed effectively is for the author to submit them and participate in the code review dialog.  While unlikely, there is a possibility that my contributions won't be reviewed as carefully because there is an assumption that I know what I'm doing, because of my long history with the project.

**Comment by davydden** [comment]
@jeffhammond 

> Creating pull requests based upon third-party patches is problematic on a number of levels.

I am sorry, I don't quite understand what you are speaking about. Those patches are NOT third-party-patches, they are official patches from NWChem, hosted on their website, i.e. http://www.nwchem-sw.org/images/Tddft_mxvec20.patch.gz . If I recall correctly, at the time this PR was added it' was not possible to download patches from urls, as I said the current state of the package simply downloads all of them from the upstream location https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/nwchem/package.py#L49-L71

**Comment by jeffhammond** [comment]
@davydden It sounds like the patches in question are already part of the NWChem development version and thus are unnecessary for supporting that in Spack.  If downloading patches is such a pain, why don't you just use the repo (svn branches for older releases; git for the latest)?

**Pull request merged**